pkg/nb/rpc_ws.go

```diff
@@ -0,0 +1,334 @@
+package nb
+
+import (
+	"context"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"nhooyr.io/websocket"
+)
+
+// RPCConnWS is an websocket connection which is shared and multiplexed
+// for all concurrent requests to the same address
+type RPCConnWS struct {
+	RPC             *RPC
+	Address         string
+	State           string
+	WS              *websocket.Conn
+	PendingRequests map[string]*RPCPendingRequest
+	NextRequestID   uint64
+	Lock            sync.Mutex
+	ReconnectDelay  time.Duration
+}
+
+// RPCPendingRequest is a struct that describes the fields related to an rpc pending requests
+type RPCPendingRequest struct {
+	Conn      *RPCConnWS
+	Req       *RPCMessage
+	Res       RPCResponse
+	ReplyChan chan error
+}
+
+// NewRPCConnWS returns a new websocket connection
+func NewRPCConnWS(r *RPC, address string) *RPCConnWS {
+	return &RPCConnWS{
+		RPC:             r,
+		Address:         address,
+		State:           "init",
+		PendingRequests: map[string]*RPCPendingRequest{},
+		Lock:            sync.Mutex{},
+	}
+}
+
+// GetAddress returns the connection address
+func (c *RPCConnWS) GetAddress() string {
+	return c.Address
+}
+
+// Reconnect connects after setting a delay
+func (c *RPCConnWS) Reconnect() {
+	c.ReconnectDelay = 3 * time.Second
+	c.Connect()
```

Comment:
We forgot to lock the connection here because Connect requires a lock...

Reply:
true, but I think the lock should be inside connect because before the dial there is a delay..